syntax = "proto3";
package spellfire;
 
import "types.proto";
 
option go_package = "github.com/Icemaster-Eric/Spellfire/backend/internal/proto;pb";
 
message Collider {
  enum ColliderType {
    COLLIDER_TYPE_UNSPECIFIED = 0;
    CIRCLE = 1;
    RECT = 2;
  }
  ColliderType type = 1;
  double rotation = 2;
  double width = 3;
  double height = 4;
  double radius = 5;
  Vec2 position = 6;
  Vec2 velocity = 7;
  bool is_static = 8;
}
 
enum Sprite {
  SPRITE_UNSPECIFIED = 0;
  SPRITE_PLAYER_GUNNER = 1;
  SPRITE_PLAYER_MAGE = 2;
  SPRITE_BUSH_1 = 3;
  SPRITE_TREE_1 = 4;
  SPRITE_TREE_2 = 5;
  SPRITE_ROCK_1 = 6;
  SPRITE_ROCK_2 = 7;
  SPRITE_ROCK_3 = 8;
  SPRITE_ROCK_4 = 9;
}
 
message RenderData { Sprite sprite = 1; }
 
enum Gun {
  GUN_UNSPECIFIED = 0;
  GUN_AUTOMATIC_RIFLE = 1;
}
 
message EntityAttribute {
  enum EntityAttributeType {
    ENTITY_ATTRIBUTE_TYPE_UNSPECIFIED = 0;
    NAME = 1;
    HEALTH = 2;
    GUN = 3;
  }
  EntityAttributeType type = 1;
  string name = 2;
  uint32 health = 3;
  Gun gun = 4;
}
 
message Entity {
  enum EntityType {
    ENTITY_TYPE_UNSPECIFIED = 0;
    GUNNER = 1;
    MAGE = 2;
  }
  enum EntityState {
    ENTITY_STATE_UNSPECIFIED = 0;
    RELOADING = 1;
  }
  uint32 id = 1;
  EntityType type = 2;
  Collider collider = 3;
  RenderData render_data = 4;
  repeated EntityState states = 5;
  repeated EntityAttribute attributes = 6;
}
 
message ServerEvent {
  enum ServerEventType {
    SERVER_EVENT_TYPE_UNSPECIFIED = 0;
    ENTER_GAME = 1;
  }
  ServerEventType type = 1;
  uint32 enter_game_player_id = 2;
}
 
message ServerPacket {
  Timestamp timestamp = 1;
  repeated Entity entities = 2;
  repeated ServerEvent events = 3;
}